### ISSUE TITLE
Installing py-llvmlite fails while trying to concretize

### ISSUE THREAD

**Issue by manuelakuhn** [LLNL/spack#24122]
I tried to install `py-llvmlite` but it fails in the concretize step

### Steps to reproduce the issue

```console
$ git clone https://github.com/spack/spack.git
$ cd spack
$ . share/spack/setup-env.sh
$ spack install py-llvmlite
```

### Error Message

```
$ spack install py-llvmlite
==> Error: An unsatisfiable variant constraint has been detected for spec:

    binutils@2.36.1%gcc@10.2.1~gas~gold~headers~interwork~ld~libiberty~lto+nls+plugins libs=shared,static arch=linux-fedora32-haswell


while trying to concretize the partial spec:

    llvm@10.0.1%gcc@10.2.1~all_targets+clang~code_signing+compiler-rt~cuda~flang+gold+internal_unwind~ipo+libcxx+lld+lldb~llvm_dylib~mlir~omp_debug~omp_tsan+polly~python~shared_libs~split_dwarf build_type=Release cuda_arch=none arch=linux-fedora32-haswell
        ^cmake@3.20.3%gcc@10.2.1~doc+ncurses+openssl+ownlibs~qt build_type=Release arch=linux-fedora32-haswell
            ^ncurses@6.2%gcc@10.2.1~symlinks+termlib abi=none arch=linux-fedora32-haswell
                ^pkgconf@1.7.4%gcc@10.2.1 arch=linux-fedora32-haswell
            ^openssl@1.1.1k%gcc@10.2.1~docs+systemcerts arch=linux-fedora32-haswell
                ^perl@5.34.0%gcc@10.2.1+cpanm+shared+threads arch=linux-fedora32-haswell
                    ^berkeley-db@18.1.40%gcc@10.2.1+cxx~docs+stl arch=linux-fedora32-haswell
                    ^gdbm@1.19%gcc@10.2.1 arch=linux-fedora32-haswell
                        ^readline@8.1%gcc@10.2.1 arch=linux-fedora32-haswell
                ^zlib@1.2.11%gcc@10.2.1+optimize+pic+shared arch=linux-fedora32-haswell
        ^hwloc@2.4.1%gcc@10.2.1~cairo~cuda~gl~libudev+libxml2~netloc~nvml+pci+shared arch=linux-fedora32-haswell
            ^libpciaccess
                ^libtool
                    ^m4@1.4.6:
                ^util-macros
            ^libxml2@2.9.10%gcc@10.2.1~python arch=linux-fedora32-haswell
                ^libiconv@1.16%gcc@10.2.1 arch=linux-fedora32-haswell
                ^xz@5.2.5%gcc@10.2.1~pic libs=shared,static arch=linux-fedora32-haswell
        ^libedit
        ^perl-data-dumper@2.173%gcc@10.2.1 arch=linux-fedora32-haswell
        ^python@3.8.10%gcc@10.2.1+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib arch=linux-fedora32-haswell
            ^bzip2@1.0.8%gcc@10.2.1~debug~pic+shared arch=linux-fedora32-haswell
                ^diffutils@3.7%gcc@10.2.1 arch=linux-fedora32-haswell
            ^expat@2.3.0%gcc@10.2.1+libbsd arch=linux-fedora32-haswell
                ^libbsd
            ^gettext@0.21%gcc@10.2.1+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-fedora32-haswell
                ^tar
            ^libffi@3.3%gcc@10.2.1 arch=linux-fedora32-haswell
            ^sqlite@3.35.5%gcc@10.2.1+column_metadata+fts~functions~rtree arch=linux-fedora32-haswell
            ^util-linux-uuid@2.36.2%gcc@10.2.1 arch=linux-fedora32-haswell
        ^swig
            ^pcre
        ^z3@4.8.9%gcc@10.2.1+python arch=linux-fedora32-haswell
            ^py-setuptools@50.3.2%gcc@10.2.1 arch=linux-fedora32-haswell


llvm requires binutils variant +gold, but spec asked for ~gold
```
Strange side note: the first time the error is about the `+gold` variant, but the second and following times it complains about the `+ld` variant. So the error message is not consistent.

### Information on your system

spack debug report
* **Spack:** 0.16.2-2988-ef9d3a464f
* **Python:** 3.8.10
* **Platform:** linux-fedora32-haswell
* **Concretizer:** original

### Additional information

```
$ spack install py-llvmlite ^binutils+gold+ld
```
seems to fix it but because I was not sure about the implications it might have I did not add it to the `py-llvmlite/package.py`

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [ ] I have run the failing commands in debug mode and reported the output

**Comment by adamjstewart** [comment]
This should be solved by Spack's new concretizer. If you add:
```yaml
config:
  concretizer: clingo
```
to your `config.yaml`, does this work? If you want to stick with the old concretizer you'll have to help it out a bit, like `spack install py-llvmlite ^binutils+gold`.

**Comment by manuelakuhn** [comment]
I wanted to try that, but unfortunately I was not able to get to the test because clingo did not build:

```
4 errors found in build log:
     3     ==> [2021-06-04-09:27:33.700729] 'cmake' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:STRING=/home/nela/.sp
           ack/bootstrap/store/linux-fedora32-x86_64/gcc-10.2.1/clingo-bootstrap-spack-4ssfscctped2mw4djzlnba4dm4ky46es
           ' '-DCMAKE_BUILD_TYPE:STRING=Release' '-DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=OFF' '-DCMAKE_VERBOSE_MAKEF
           ILE:BOOL=ON' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=OFF' '-DCMAKE_INSTALL_RPATH:STRING=/home/nela/.spack/
           bootstrap/store/linux-fedora32-x86_64/gcc-10.2.1/clingo-bootstrap-spack-4ssfscctped2mw4djzlnba4dm4ky46es/lib
           ;/home/nela/.spack/bootstrap/store/linux-fedora32-x86_64/gcc-10.2.1/clingo-bootstrap-spack-4ssfscctped2mw4dj
           zlnba4dm4ky46es/lib64' '-DCMAKE_PREFIX_PATH:STRING=/home/nela/.spack/bootstrap/store/linux-fedora32-x86_64/g
           cc-10.2.1/bison-3.7.6-vou46c57zizqhle6kktywol7zumcaxnj;/home/nela/.spack/bootstrap/store/linux-fedora32-x86_
           64/gcc-10.2.1/cmake-3.20.3-zhmxauwzvssu3pfxrrhoxq2x7z2w2c7z;/home/nela/.spack/bootstrap/store/linux-fedora32
           -x86_64/gcc-10.2.1/re2c-1.2.1-sqdifw5yv5clj7lp6bsgrp2re7cfewf7' '-DCLINGO_REQUIRE_PYTHON=ON' '-DCLINGO_BUILD
           _WITH_PYTHON=ON' '-DPYCLINGO_USER_INSTALL=OFF' '-DPYCLINGO_USE_INSTALL_PREFIX=ON' '-DCLINGO_BUILD_WITH_LUA=O
           FF' '-DCLINGO_BUILD_PY_SHARED:STRING=OFF' '-DPython_EXECUTABLE:STRING=/usr/bin/python3.8' '-DPython_INCLUDE_
           DIR:STRING=/usr/include/python3.8' '-DCLINGO_BUILD_APPS:STRING=OFF' '/tmp/nela/spack-stage/spack-stage-cling
           o-bootstrap-spack-4ssfscctped2mw4djzlnba4dm4ky46es/spack-src'
     4     -- The CXX compiler identification is GNU 10.3.1
     5     -- Detecting CXX compiler ABI info
     6     -- Detecting CXX compiler ABI info - failed
     7     -- Check for working CXX compiler: /tmp/spack_test/lib/spack/env/gcc/g++
     8     -- Check for working CXX compiler: /tmp/spack_test/lib/spack/env/gcc/g++ - broken
  >> 9     CMake Error at /home/nela/.spack/bootstrap/store/linux-fedora32-x86_64/gcc-10.2.1/cmake-3.20.3-zhmxauwzvssu3
           pfxrrhoxq2x7z2w2c7z/share/cmake-3.20/Modules/CMakeTestCXXCompiler.cmake:59 (message):
     10      The C++ compiler
     11    
     12        "/tmp/spack_test/lib/spack/env/gcc/g++"
     13    
     14      is not able to compile a simple test program.
     15    

     ...

     21        gmake[1]: Entering directory '/tmp/nela/spack-stage/spack-stage-clingo-bootstrap-spack-4ssfscctped2mw4dj
           zlnba4dm4ky46es/spack-build-4ssfscc/CMakeFiles/CMakeTmp'
     22        Building CXX object CMakeFiles/cmTC_4a3e9.dir/testCXXCompiler.cxx.o
     23        /tmp/spack_test/lib/spack/env/gcc/g++   -static-libstdc++ -static-libgcc -Wl,--exclude-libs,ALL  -o CMak
           eFiles/cmTC_4a3e9.dir/testCXXCompiler.cxx.o -c /tmp/nela/spack-stage/spack-stage-clingo-bootstrap-spack-4ssf
           scctped2mw4djzlnba4dm4ky46es/spack-build-4ssfscc/CMakeFiles/CMakeTmp/testCXXCompiler.cxx
     24        Linking CXX executable cmTC_4a3e9
     25        /home/nela/.spack/bootstrap/store/linux-fedora32-x86_64/gcc-10.2.1/cmake-3.20.3-zhmxauwzvssu3pfxrrhoxq2x
           7z2w2c7z/bin/cmake -E cmake_link_script CMakeFiles/cmTC_4a3e9.dir/link.txt --verbose=1
     26        /tmp/spack_test/lib/spack/env/gcc/g++ -static-libstdc++ -static-libgcc -Wl,--exclude-libs,ALL  -static-l
           ibstdc++ -static-libgcc -Wl,--exclude-libs,ALL  -rdynamic CMakeFiles/cmTC_4a3e9.dir/testCXXCompiler.cxx.o -o
            cmTC_4a3e9
  >> 27    /usr/bin/ld: cannot find -lstdc++
  >> 28        collect2: error: ld returned 1 exit status
  >> 29        gmake[1]: *** [CMakeFiles/cmTC_4a3e9.dir/build.make:99: cmTC_4a3e9] Error 1
     30        gmake[1]: Leaving directory '/tmp/nela/spack-stage/spack-stage-clingo-bootstrap-spack-4ssfscctped2mw4djz
           lnba4dm4ky46es/spack-build-4ssfscc/CMakeFiles/CMakeTmp'
     31        gmake: *** [Makefile:127: cmTC_4a3e9/fast] Error 2
     32    
     33    
     34    
     35  
```
I also just tried that with a new Spack: 0.16.2-3007-ea390198f4, same result

In case that matters: python3 headers and libstdc++ are installed on my system. (installed packages: python3-devel, libstdc++, libstdc++-devel)

**Comment by alalazo** [comment]
@manuelakuhn Can you install static libstdc++ ? To bootstrap from sources we require that at the moment.

**Comment by manuelakuhn** [comment]
Sorry for the delayed reply, I wanted to make sure that it actually builds. On the test system from above, llvm takes ages to build and then fails because I did not pay attention to set the build_stage to a reasonable path on that system.
Unfortunately on the other test system it did not build because llvm@10 (requirement from py-llvmlite) does not build with gcc@11. Since this is a different topic, I can open a separate issue for it.
So to summarize: yes clingo works for this.

In general: for testing to switch to clingo would be ok for me, but for my productive setup, I do not want to use it yet. 
The reason I opened a ticket originally was that because `py-llvmlite` is broken with the default concretizer also all packages depending on it are. E.g I stumbled in it because I wanted to create a new package with a dependency on `py-numba` which has a dependency on `py-llvmlite` again. That means the whole chain is broken with the default concretizer if `py-llvmlite` is not patched. I guess other packages are affected as well.
